napari/layers/layergroup/layergroup.py

```diff
@@ -0,0 +1,289 @@
+from __future__ import annotations
+
+import itertools
+import warnings
+from typing import TYPE_CHECKING, Iterable, List, Optional
+
+import numpy as np
+
+from ...components.layerlist import Extent, _LayerListMixin
+from ...utils.context import create_context
+from ...utils.context._layerlist_context import LayerListContextKeys
+from ...utils.naming import inc_name_count
+from ...utils.translations import trans
+from ...utils.tree import Group
+from ..base import Layer
+from ..utils.layer_utils import combine_extents
+
+if TYPE_CHECKING:
+    from npe2.manifest.io import WriterContribution
+
+
+class LayerGroup(Group[Layer], Layer, _LayerListMixin):
```

Comment:
`LayerGroup` sub-typing `Layer` makes me uncomfortable (for reasons I'm struggling to coherently describe), but my guess it's almost required to make progress here - is that right?

I think it's because I think of `Layer.data` as a VIP in napari, but `LayerGroup` is a `Layer` that I assume we expect will always have no data. I think this kind of mirrors the separation of Node/Group as Leaf/Composition, in that we only expect leaf nodes in the layer group tree to have data.

Reply:
yep, this is the composite pattern that @jni proposed a while back and we all decided we liked.  This is the thing that makes it possible for vispy, for example, to _only_ interact with the layergroup (which does the work of deferring to it's components).  It's pairs perfectly with the scenegraph idea, where the top level scene need only pay attention to top level nodes, and needn't recurse all the way down. 

I feel your hesitation.  so feel free to keep contemplating that and poking at it.  So far, that particular part hasn't been problematic, and is pretty well covered by tests here.  but definitely open to thoughts